idb/typeinf.py

```diff
@@ -0,0 +1,1713 @@
+                member.fda = attr
+                if not member.is_baseclass():
+                    if len(fields) > field_i:
+                        member.name = fields[field_i]
+                    if n < len(fieldcmts):
+                        member.cmt = fieldcmts[field_i]
+                    field_i += 1
+                self.members.append(member)
+        return self
+
+
+TAENUM_64BIT = 0x0020  # enum: store 64-bit values
+
+
+class EnumMember:
+    def __init__(self, name, value=0, cmt=""):
+        self.name = name  # qstring
+        self.cmt = cmt  # qstring
+        self.value = value
+
+
+def to_s32(n):
+    n = n & 0xFFFFFFFF
+    return n | (-(n & 0x80000000))
+
+
+class EnumTypeData(TypeData):
+    """Representation of enum_type_data_t"""
+
+    def __init__(self):
+        TypeData.__init__(self)
+        self.group_sizes = []  # intvec_t
+        # TAENUM_64BIT   0x0020
+        #  	enum: store 64-bit values
+        self.taenum_bits = 0
+        self.bte = 0
+        self.members = []
+
+    def deserialize(self, til, ts, fields, fieldcmts):
+        typ = ts.u8()
+        n = ts.complex_n()
+        if n == 0:
+            ts.pstring()
+            ts.tah_attr()
+        else:
+            self.taenum_bits = ts.tah_attr()
+            self.bte = ts.u8()
+            cur = 0
+            hi = 0
+            mask = self.calc_mask(til)
+            for i in range(n):
+                # TODO: subarrays
+                # https://www.hex-rays.com/products/ida/support/sdkdoc/group__tf__enum.html#ga9ae7aa54dbc597ec17cbb17555306a02
+                if self.taenum_bits & TAENUM_64BIT:
+                    hi = ts.de()
+                if self.bte & BTE_BITFIELD:
+                    self.group_sizes.append(ts.dt())
+                lo = ts.de()
+                cur += to_s32(lo | (hi << 32) & mask)
+                member = EnumMember(fields[i], value=cur)
+                self.members.append(member)
+        return self
+
+    def calc_mask(self, til):
+        emsize = self.bte & BTE_SIZE_MASK
+        if emsize != 0:
+            bytesize = 1 << (emsize - 1)
+        else:
+            bytesize = til.size_e
+        bitsize = bytesize * 8
+        if bitsize < 64:
+            return (1 << bitsize) - 1
+        return 0xFFFFFFFFFFFFFFFF
+
+
+class TypedefTypeData(TypeData):
+    """Representation of typedef_type_data_t"""
+
+    def __init__(self):
+        TypeData.__init__(self)
+        self.til = None
+        # union
+        # {
+        #    const char* name is_ordref=false: target type name. we do not own this pointer!
+        #    uint32   ordinal is_ordref=true: type ordinal number
+        # }
+        self.name = None
+        self.ordinal = 0
+        self.is_ordref = False
+        self.resolve = False
+
+    def deserialize(self, til, type_string, fields, fieldcmts):
+        self.til = til
+        typ = type_string.u8()
+        buf = type_string.pbytes()
+        if len(buf) > 1 and buf.find(b"#") == 0:
```

Comment:
```
buf.find(b"#")
```
Won't that search the entire string for the "#" character? It should only be checking the first character. The format of an ordinal is a "#" character then a `de()` which when decoded is the type ordinal (See `is_ordinal_name()`).

Reply:
`buf.startswith(b"#")`

seems good